packages/realm-react/src/__tests__/useEmailPasswordAuth.test.tsx

```diff
@@ -0,0 +1,241 @@
+      });
+
+      // Get an instance of the realm app and make sure the current user has been set
+      const realmApp = new App({ id: appId, baseUrl });
+      expect(realmApp.currentUser).not.toBeNull();
+      await act(async () => {
+        await result.current.logOut();
+      });
+      expect(realmApp.currentUser).toBeNull();
+    });
+    it("can register and login with email/password, just by calling register and login", async () => {
+      const { result } = renderEmailPasswordAuth(appId, baseUrl);
+      await act(async () => {
+        result.current.register({ email: "test2@test.com", password: "password", loginAfterRegister: false });
+        await waitFor(() => {
+          expect(result.current.result.pending).toEqual(true);
+        });
+      });
+      await waitFor(() => {
+        expect(result.current.result.success).toEqual(true);
+      });
+
+      await act(async () => {
+        result.current.logIn({ email: "test2@test.com", password: "password" });
+        await waitFor(() => {
+          expect(result.current.result.pending).toEqual(true);
+        });
+      });
+      await waitFor(() => {
+        expect(result.current.result.success).toEqual(true);
+      });
+
+      // Get an instance of the realm app and make sure the current user has been set
+      const realmApp = new App({ id: appId, baseUrl });
+      const user = realmApp.currentUser;
+      expect(user).not.toBeNull();
+
+      await act(async () => {
+        await result.current.logOut();
+      });
+      expect(realmApp.currentUser).toBeNull();
+    });
+    it("sets an error state when user is already registered", async () => {
+      const { result } = renderEmailPasswordAuth(appId, baseUrl);
+
+      await act(async () => {
+        result.current.register({ email: "test2@test.com", password: "password", loginAfterRegister: false }),
+          await waitFor(() => {
+            expect(result.current.result.pending).toEqual(true);
+          });
+      });
+      await waitFor(() => {
+        expect(result.current.result.state).toBe(OperationState.Error);
+        expect(result.current.result.error).toBeDefined();
+      });
+    });
+  });
+  describe("all methods are callable and report a state", () => {
+    let appId: string;
+    beforeAll(async () => {
+      const config = new AppConfigBuilder("test-app-2");
+      config.authProvider({
+        name: "local-userpass",
+        type: "local-userpass",
+        config: {
+          autoConfirm: true,
+          resetPasswordUrl: "http://localhost/resetPassword",
+        },
+        disabled: false,
+      });
+      ({ appId } = await importApp(config.config));
+    });
+    it("logIn", async () => {
+      const { result } = renderEmailPasswordAuth(appId, baseUrl);
+      await act(async () => {
+        result.current.logIn({ email: "test@test.com", password: "password" });
+        await waitFor(() => {
+          expect(result.current.result.pending).toBeTruthy();
```

Comment:
I suspect there might be a race condition here, if the error happens so fast that it never sees the `pending` state?

Reply:
Can you expand on this?  Where would the race condition be?  In the test, or the lib?

Reply:
In the test. `waitFor` will call the callback and poll until the callback no longer throws, but if the network is fast, it might go straight into the `error` state before it gets into `pending`, which would leave the test hanging indefinitely.

Reply:
It's set to `pending` right before the operation is invoked, so I doubt it would be possible.  I could be wrong though.  In any case, for the most part we should be waiting for the result.  Although I do want to verify in the tests that `pending` is set to true after the method is invoked.  I guess this could be done though mocking rather than assuming 😂

Reply:
> It's set to pending right before the operation is invoked

Even so, it could still fail and switch to error before the callback passed to `waitFor` is invoked (depending on the implementation of `waitFor`).